File: quotation-service/src/auth/jwt.ts
import jwt from 'jsonwebtoken';
import type { UserRole } from '../db.js';

export interface AuthTokenPayload {
  userId: number;
  email: string;
  role: UserRole;
}

const JWT_EXPIRES_IN = '12h';

function getSecret(): string {
  const secret = process.env.JWT_SECRET;
  if (!secret || secret.trim() === '') {
    throw new Error('JWT_SECRET n\'est pas configuré (voir quotation-service/.env.example).');
  }
  return secret;
}

export function signAuthToken(payload: AuthTokenPayload): string {
  return jwt.sign(payload, getSecret(), { expiresIn: JWT_EXPIRES_IN });
}

/** Returns the decoded payload, or null if the token is missing/invalid/expired. */
export function verifyAuthToken(token: string): AuthTokenPayload | null {
  try {
    return jwt.verify(token, getSecret()) as unknown as AuthTokenPayload;
  } catch {
    return null;
  }
}

File: vector-service/app/auth/jwt.py
from datetime import datetime, timedelta, timezone
from typing import Optional

from jose import JWTError, jwt

from .. import config

ALGORITHM = "HS256"
EXPIRES_DELTA = timedelta(hours=12)


def sign_auth_token(user_id: int, email: str, role: str) -> str:
    if not config.JWT_SECRET:
        raise RuntimeError("JWT_SECRET n'est pas configuré (voir vector-service/.env.example).")
    payload = {
        "userId": user_id,
        "email": email,
        "role": role,
        "exp": datetime.now(timezone.utc) + EXPIRES_DELTA,
    }
    return jwt.encode(payload, config.JWT_SECRET, algorithm=ALGORITHM)


def verify_auth_token(token: str) -> Optional[dict]:
    if not config.JWT_SECRET:
        return None
    try:
        return jwt.decode(token, config.JWT_SECRET, algorithms=[ALGORITHM])
    except JWTError:
        return None
